eng/tools/GenerateFiles/Directory.Build.targets.in

```diff
@@ -2,43 +2,65 @@
+        version used in most projects. Projects can use $(TargetLatestRuntimePatch) to explicitly control whether
+        assemblies build against default (false) or latest (true). When that property is not set, SDK uses
+        default metadata in most cases but published apps e.g. tool projects (again, property not set) use latest.
+      -->
+      <DefaultRuntimeFrameworkVersion
+          Condition=" '$(IsServicingBuild)' != 'true' ">${MicrosoftNETCoreAppRuntimeVersion}</DefaultRuntimeFrameworkVersion>
+      <TargetingPackVersion
+          Condition=" '$(IsServicingBuild)' != 'true' ">${MicrosoftNETCoreAppRefVersion}</TargetingPackVersion>
+    </KnownFrameworkReference>
 
-    <!-- Use the just-built ASP.NET Core shared framework. -->
+    <!-- Use the just-built ASP.NET Core shared framework if available except when building product code in servicing. -->
     <KnownFrameworkReference
-      Update="@(KnownFrameworkReference->WithMetadataValue('Identity', 'Microsoft.AspNetCore.App')->WithMetadataValue('TargetFramework', '${KnownAppHostPackOrFrameworkReferenceTfm}'))"
-      DefaultRuntimeFrameworkVersion="${MicrosoftAspNetCoreAppRuntimeVersion}"
-      LatestRuntimeFrameworkVersion="${MicrosoftAspNetCoreAppRuntimeVersion}"
-      TargetingPackVersion="${MicrosoftAspNetCoreAppRefVersion}"
-      RuntimePackRuntimeIdentifiers="${SupportedRuntimeIdentifiers}" />
-  </ItemGroup>
+        Update="@(KnownFrameworkReference->WithMetadataValue('Identity', 'Microsoft.AspNetCore.App')->WithMetadataValue('TargetFramework', '${KnownAppHostPackOrFrameworkReferenceTfm}'))"
+        Condition=" $(UpdateAspNetCoreKnownFramework) "
+        LatestRuntimeFrameworkVersion="${MicrosoftAspNetCoreAppRuntimeVersion}"
+        RuntimePackRuntimeIdentifiers="${SupportedRuntimeIdentifiers}">
+      <DefaultRuntimeFrameworkVersion
+          Condition=" '$(IsServicingBuild)' != 'true' ">${MicrosoftAspNetCoreAppRuntimeVersion}</DefaultRuntimeFrameworkVersion>
+      <TargetingPackVersion
+          Condition=" '$(IsServicingBuild)' != 'true' ">${MicrosoftAspNetCoreAppRefVersion}</TargetingPackVersion>
+    </KnownFrameworkReference>
 
-  <ItemGroup Condition="$(BuildingTestAppsIndependently) == 'false'">
     <PackageReference Include="Microsoft.Net.Compilers.Toolset"
       Version="${MicrosoftNetCompilersToolsetVersion}"
       PrivateAssets="all"
       IsImplicitlyDefined="true" />
   </ItemGroup>
+
+  <!-- Warn if the "just-built" ASP.NET Core shared framework does not exist. -->
+  <Target Name="WarnAboutMissingSharedFramework"
+      BeforeTargets="Restore;Build;Rebuild;RunTests;Test;VSTest;Pack"
+      Condition=" $(UpdateAspNetCoreKnownFramework) ">
+    <Warning Text="Requested Microsoft.AspNetCore.App v${MicrosoftAspNetCoreAppRuntimeVersion} does not exist."
+        Condition="!EXISTS('$(DOTNET_ROOT)\shared\Microsoft.AspNetCore.App\${MicrosoftAspNetCoreAppRuntimeVersion}')" />
```

Comment:
@rainersigwald we hit this warning in the `RestoreTestProjects` targets in src/Components/WebAssembly/Sdk/integrationtests/Microsoft.NET.Sdk.BlazorWebAssembly.IntegrationTests.csproj and src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Microsoft.NET.Sdk.Razor.IntegrationTests.csproj though those targets execute a while after the src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj has laid out the newly-built shared framework in this directory. Is the `EXISTS` check using a cache of the directory content from earlier in the build❔ If yes, what's the workaround (I prefer to avoid separating those test projects into another build step)❔ If no, what am I doing wrong❔

Reply:
Exists as a condition inside a target should happen late and not be cached. I'm not aware of any bugs in that area.

How are you determining the order of operations here? Is there a binlog uploaded in the build?

Reply:
I reproduced the problem in a local build and used the timeline view in that binary log. Because this requires a clean repo, including removal of 6.0.0-dev folders under .dotnet/, the binary log is pretty large❕ But, I'll do it again and share the results this afternoon&hellip;

Reply:
Bit more detail:
``` text
> dotnet msbuild .\artifacts\log\Debug\Build.binlog /noconlog '/flp:v=diag;logfile=diag.log'
...
> sls '-\>.*\\.dotnet\\shared', 'error :' .\diag.log
diag.log:39258399:                     Task Parameter:Text=Microsoft.AspNetCore.App.Runtime.csproj -> ...\.dotnet\shared\Microsoft.AspNetCore.App\6.0.0-dev\ (TaskId:7247)
diag.log:39258400:                     Microsoft.AspNetCore.App.Runtime.csproj -> ...\.dotnet\shared\Microsoft.AspNetCore.App\6.0.0-dev\ (TaskId:7247)
diag.log:57244108:19:33:52.124   529>...\artifacts\bin\GenerateFiles\Directory.Build.targets(74,5): error : Requested Microsoft.AspNetCore.App v6.0.0-dev does not exist. [...\src\Components\WebAssembly\Sdk\testassets\RestoreBlazorWasmTestProjects\RestoreBlazorWasmTestProjects.csproj]
diag.log:57244113:19:33:52.124   529>...\artifacts\bin\GenerateFiles\Directory.Build.targets(74,5): error : Requested Microsoft.AspNetCore.App v6.0.0-dev does not exist. [...\src\Components\WebAssembly\Sdk\testassets\RestoreBlazorWasmTestProjects\RestoreBlazorWasmTestProjects.csproj]
diag.log:65080699:19:37:05.409   535>...\artifacts\bin\GenerateFiles\Directory.Build.targets(74,5): error : Requested Microsoft.AspNetCore.App v6.0.0-dev does not exist.
...
> . .\activate.ps1
Enabled the .NET Core environment. Execute 'deactivate' to exit.
dotnet = C:\dd\dnx\aspnetcore\.dotnet\dotnet.exe
> dir "$env:DOTNET_ROOT\shared\Microsoft.AspNetCore.App\6.0.0-dev\"
    Directory: ...\.dotnet\shared\Microsoft.AspNetCore.App\6.0.0-dev

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          2020-11-17    19:30             53 .version
-a---          2020-11-17    19:13        1282048 aspnetcorev2_inprocess.dll
-a---          2020-11-17    19:25          48640 Microsoft.AspNetCore.Antiforgery.dll
-a---          2020-11-17    19:30          21393 Microsoft.AspNetCore.App.deps.json
...
```

Reply:
The 6.0.0-dev directory was last updated at 19:30:38, a few minutes before the problems started at 19:33:52.

Reply:
@rainersigwald ignore all of the above. The issue was my incorrect assumption `$(DOTNET_ROOT)` was always set. Noticing the problem went away in _some_ of my local builds and some time up close and personal with binary logs helped (eventually).